README.md
<!-- -*- mode:gfm -*- github-flavored markdown -->

# certbot-dns-joker

Certbot DNS Authenticator plugin for [Joker](https://joker.com/).

This [Certbot](https://certbot.eff.org/docs/index.html) plugin automates the
process of completing a `dns-01` by creating (and removing) TXT records using
the [Joker TXT record
API](https://joker.com/faq/content/6/496/en/let_s-encrypt-support.html).

## Configuring Joker

For each of your domains hosted using the Joker DNS server that would would
like to obtain certificates for you must enable Dynamic DNS for the domain in
the Joker web console.  Do this by visiting your [Joker
Dashboard](https://joker.com/user/dashboard), clicking the "DNS" action for
the domain you want to enable Dynamic DNS for, then ensuring that the "Dynamic
DNS active" slider is turned on.  A dialog should appear with the DynDNS
username and password for that domain.  These will be used in the credentials
file described below.

## Installation

``` bash
pip install certbot-dns-joker
```

## Certbot Arguments

To use Joker DNS authentication, pass the following arguments on certbot's command line:

| Option | Description |
| --- | --- |
| `--authenticator dns-joker` | Select the Joker authenticator plugin. (required) |
| `--dns-joker-credentials` _credentials_file_ | Full path to config file containing domain credentials. |
| `--dns-joker-propagation-seconds` _delay_ | Delay between setting DNS TXT record and asking the ACME server to verify it. Default: 120 |

If you don't supply the credentials file on the certbot command line you will
be prompted for its location.

## Credentials

You need to create a configuration file on your system (for example
`/etc/letsencrypt/secrets/DOMAIN.ini`) that contains the per-domain secrets
that you obtained when you enabled DynDNS for your domain.

``` plain
dns_joker_username = USERNAME
dns_joker_password = PASSWORD
dns_joker_domain = DOMAIN
```

## Example

``` bash
certbot certonly \
  --authenticator dns-joker \
  --dns-joker-credentials /etc/letsencrypt/secrets/example.com.ini \
  -d example.com -d '*.example.com'
```

## Alternate Deployment Methods

### PyPI

This project has a PyPI page at <https://pypi.org/project/certbot-dns-joker/>.

### RPM

I have built an RPM for CentOS 8, which is the platform I use.  The Makefile
has an `rpm` target that builds the RPM, but this has only been tested on
CentOS 8.  In addition, the GitHub release contains an SRPM from which is
should be possible to build a new RPM on other RPM-based Linux distributions,
but again I have not tested this.

### Docker

The Makefile has a `docker-image` target to create a certbot docker image with
the certbot-dns-joker plugin installed.  In addition if you set the
DOCKER_OTHER_PLUGINS makefile variable to a (space-separated) list of plugins
those plugins will also be installed.

Once the docker image is built you can run it with a command such as the
following.

``` bash
docker run --rm \
  -v /var/lib/letsencrypt:/var/lib/letsencrypt \
  -v /etc/letsencrypt:/etc/letsencrypt \
  -v /var/log/letsencrypt:/var/log/letsencrypt \
  --cap-drop=all \
  certbot-joker \
  certonly \
  --authenticator dns-joker \
  --dns-joker-propagation-seconds 900 \
  --dns-joker-credentials /etc/letsencrypt/secrets/example.com.ini \
  --no-self-upgrade \
  --keep-until-expiring --non-interactive --expand \
  --server https://acme-v02.api.letsencrypt.org/directory \
  -d example.com -d '*.example.com'
```

Then you can run a command such as the following from cron to renew your
certificates.

``` bash
docker run --rm \
  -v /var/lib/letsencrypt:/var/lib/letsencrypt \
  -v /etc/letsencrypt:/etc/letsencrypt \
  -v /var/log/letsencrypt:/var/log/letsencrypt \
  --cap-drop=all \
  certbot-joker \
  renew
```

Note that plugins that attempt to do operations outside of the container (such
as the apache plugin, which wants to run apachectl) will fail.

## Upgrading from 1.2.0

If you have existing conf files in `/etc/letsencrypt/renewal`, you will need
to edit them to remove `certbot-dns-joker:` from the `authenticator` and
credentials lines.

``` bash
perl -p -i.bak -e 's/certbot-dns-joker://;' /etc/letsencrypt/renewal/*.conf
```

## Acknowledgments

This plugin is based in large part on Matthias Bilger's
[certbot-dns-ispconfig](https://github.com/m42e/certbot-dns-ispconfig) plugin
and the certbot-dns-dnsimple plugin.


CHANGELOG.md
# Changelog

## Version 2.1.0 &mdash; 2023-02-15

This is version 2.1.0 of certbot-dns-joker because the convention is now that
a Certbot plugin uses the same version number as the Certbot version it is
built against.  There are no major changes in this version of
certbot-dns-joker.

### New Features

* Updates for Certbot 2.1.0.
* As a Certbot 2.0.0 change, config variables and names no longer use the
  `certbot-dns-joker:` prefix.  If upgrading, see the [Upgrading from
  1.2.0](https://github.com/dhull/certbot-dns-joker/README.md#upgrading-from-120)
  section of the README.

## Version 1.1.0 &mdash; 2020-06-23

Previously certbot-dns-joker only allowed getting certificates at the root
level of the domain, so, for example, if you owned the domain "example.com"
you could get a certificate for "example.com" or "*.example.com".  Now you can
also get a certificate for a subdomain such as sub.example.com.  Since
certbot-dns-joker doesn't know how to determine what the root domain is for a
given subdomain you will have to specify the root domain in the credentials
file as "domain = ROOT_DOMAIN".

### New Features

* Allow getting certificates for subdomains.

## Version 1.0.0 &mdash; 2020-06-21

This is the initial public release.

### New Features

* Package as wheel (whl) and build Docker image.
* Add RPM packaging.
* Publish to PyPI.
* Create GitHub releases.


tests/dns_joker_test.py
"""Tests for certbot_dns_joker.dns_joker."""

import unittest

try:
    import mock
except ImportError: # pragma: no cover
    from unittest import mock # type: ignore
from requests.exceptions import HTTPError
import urllib.parse
import requests
import requests_mock

from certbot.compat import os
from certbot.errors import PluginError
from certbot.plugins import dns_test_common
from certbot.plugins.dns_test_common import DOMAIN
from certbot.tests import util as test_util

FAKE_USERNAME = 'fake_username'
FAKE_PASSWORD = 'fake_password'
MOCK_ENDPOINT = 'mock://endpoint'

class AuthenticatorTest(test_util.TempDirTestCase,
                        dns_test_common.BaseAuthenticatorTest):

    def setUp(self):
        super(AuthenticatorTest, self).setUp()

        from certbot_dns_joker.dns_joker import Authenticator

        path = os.path.join(self.tempdir, 'file.ini')
        dns_test_common.write({
            # 'certbot_dns_joker:dns_joker_username': FAKE_USERNAME,
            # 'certbot_dns_joker:dns_joker_password': FAKE_PASSWORD,
            'joker_username': FAKE_USERNAME,
            'joker_password': FAKE_PASSWORD,
        }, path)

        self.config = mock.MagicMock(joker_credentials=path,
                                     joker_propagation_seconds=0)  # don't wait during tests

        # self.auth = Authenticator(self.config, "certbot_dns_joker:dns_joker")
        self.auth = Authenticator(self.config, "joker")

        self.mock_client = mock.MagicMock()
        # _get_joker_client | pylint: disable=protected-access
        self.auth._get_joker_client = mock.MagicMock(return_value=self.mock_client)


    # Commented out this test because it started producing the error,
    # "ValueError: This function was called too early in Certbot's execution
    # as the display utility hasn't been configured yet."  trying to call
    # `display_util.notify("Waiting %d seconds for DNS changes to propagate"
    # %`.

    # def test_perform(self):
    #     self.auth.perform([self.achall])

    #     expected = [
    #         mock.call.add_txt_record(
    #             DOMAIN, "_acme-challenge." + DOMAIN, mock.ANY
    #         )
    #     ]
    #     self.assertEqual(expected, self.mock_client.mock_calls)

    def test_cleanup(self):
        # _attempt_cleanup | pylint: disable=protected-access
        self.auth._attempt_cleanup = True
        self.auth.cleanup([self.achall])

        expected = [
            mock.call.del_txt_record(
                DOMAIN, "_acme-challenge." + DOMAIN, mock.ANY
            )
        ]
        self.assertEqual(expected, self.mock_client.mock_calls)


class JokerClientTest(unittest.TestCase):
    record_name = "_acme-challenge." + DOMAIN
    record_content = "bar"
    record_ttl = 42

    def setUp(self):
        from certbot_dns_joker.dns_joker import _JokerClient

        self.client = _JokerClient(FAKE_USERNAME, FAKE_PASSWORD, DOMAIN,
                                   self.record_ttl, endpoint=MOCK_ENDPOINT)

        self.adapter = requests_mock.Adapter()
        self.client.session.mount('mock://', self.adapter)

    def _register_response(self, response='good', subdomain=None, additional_matcher=None, **kwargs):
        def add_matcher(request):
            data = urllib.parse.parse_qs(request.text)
            add_result = True
            if additional_matcher is not None:
                add_result = additional_matcher(request)

            def submatch(label):
                if subdomain:
                    print(f'checking label:{label} subdomain:{subdomain}')
                    return len(label) > len(subdomain) and label[-len(subdomain)-1:] == '.' + subdomain
                else:
                    return True

            # The error message is unhelpful (NoMockAddress) if this fails.
            return (
                ("username" in data and data["username"] == [FAKE_USERNAME]) and
                ("password" in data and data["password"] == [FAKE_PASSWORD]) and
                ("zone" in data and data["zone"] == [DOMAIN]) and
                ("label" in data and submatch(data["label"][0])) and
                add_result
            )

        self.adapter.register_uri(
            requests_mock.ANY,
            MOCK_ENDPOINT,
            text=response,
            status_code=200 if response == 'good' else 400,
            additional_matcher=add_matcher,
            **kwargs
        )

    def test_add_txt_record(self):
        self._register_response()
        self.client.add_txt_record(
            DOMAIN, self.record_name, self.record_content
        )

    def test_add_txt_record_fail_to_authenticate(self):
        self._register_response(response='badauth')
        with self.assertRaises(PluginError) as context:
            self.client.add_txt_record(
                DOMAIN, self.record_name, self.record_content
            )


    def test_add_txt_record_fail_to_find_domain(self):
        self._register_response(response='nohost')
        with self.assertRaises(PluginError) as context:
            self.client.add_txt_record(
                DOMAIN, self.record_name, self.record_content
            )

    def test_add_txt_record_subdomain(self):
        self._register_response(subdomain='sub')
        self.client.add_txt_record(
            'sub.' + DOMAIN, 'challenge.sub.' + DOMAIN, self.record_content
        )

    def test_del_txt_record(self):
        self._register_response()
        self.client.del_txt_record(
            DOMAIN, self.record_name, self.record_content
        )


if __name__ == "__main__":
    unittest.main()  # pragma: no cover


src/certbot_dns_joker/__init__.py
"""
The `~certbot_dns_joker.dns_joker` plugin automates the process of
completing a ``dns-01`` challenge (`~acme.challenges.DNS01`) by creating, and
subsequently removing, TXT records using the Joker DynDNS API.


Named Arguments
---------------

========================================  =====================================
``--dns-joker-credentials``               Joker credentials_ INI file.
                                          (Required)
``--dns-joker-propagation-seconds``       The number of seconds to wait for DNS
                                          to propagate before asking the ACME
                                          server to verify the DNS record.
                                          (Default: 120)
========================================  =====================================


Credentials
-----------

Use of this plugin requires a configuration file containing Joker DynDNS API
credentials, obtained from your Joker account.
`https://joker.com/faq/content/6/496/en/let_s-encrypt-support.html`_.

https://configobj.readthedocs.io/en/latest/configobj.html

.. code-block:: ini
   :name: credentials.ini
   :caption: Example credentials file:

   # Joker DynDNS API credentials used by Certbot
   username = USERNAME
   password = PASSWORD

The path to this file can be provided interactively or using the
``--dns-joker-credentials`` command-line argument. Certbot records the path
to this file for use during renewal, but does not store the file's contents.

.. caution::
   You should protect these API credentials as you would the password to your
   Joker account. Users who can read this file can use these credentials
   to issue arbitrary API calls on your behalf. Users who can cause Certbot to
   run using these credentials can complete a ``dns-01`` challenge to acquire
   new certificates or revoke existing certificates for associated domains,
   even if those domains aren't being managed by this server.

Certbot will emit a warning if it detects that the credentials file can be
accessed by other users on your system. The warning reads "Unsafe permissions
on credentials configuration file", followed by the path to the credentials
file. This warning will be emitted each time Certbot uses the credentials file,
including for renewal, and cannot be silenced except by addressing the issue
(e.g., by using a command like ``chmod 600`` to restrict access to the file).


Examples
--------

.. code-block:: bash
   :caption: To acquire a certificate for ``example.com``

   certbot certonly \\
     --authenticator dns-joker \\
     --dns-joker-credentials ~/.secrets/certbot/example.com.ini \\
     -d example.com

.. code-block:: bash
   :caption: To acquire a single certificate for both ``example.com`` and
             ``www.example.com``

   certbot certonly \\
     --authenticator dns-joker \\
     --dns-joker-credentials ~/.secrets/certbot/example.com.ini \\
     -d example.com \\
     -d www.example.com

.. code-block:: bash
   :caption: To acquire a certificate for ``example.com``, waiting 60 seconds
             for DNS propagation

   certbot certonly \\
     --authenticator dns-joker \\
     --dns-joker-credentials ~/.secrets/certbot/example.com.ini \\
     --dns-joker-propagation-seconds 60 \\
     -d example.com

"""

# The following lines are extracted so they must be of the form:
# __KEY__ = "VALUE"

__version__ = "2.1.0"

__title__ = "certbot-dns-joker"
__description__ = "Joker DNS Authenticator plugin for Certbot"
__url__ = "https://github.com/dhull/certbot-dns-joker"
__doc__ = __description__ + " <" + __url__ + ">"

__author__ = "David Hull"
__email__ = "github@davidhull.org"
__issue_tracker__ = 'https://github.com/dhull/certbot-dns-joker/issues'

__license__ = "Apache 2.0"
__copyright__ = "Copyright 2022 David Hull"


src/certbot_dns_joker/dns_joker.py
"""DNS Authenticator for DNS servers with the Joker extension to the DynDNS API."""
import logging

import requests

from certbot import errors
from certbot import interfaces
from certbot.plugins import dns_common

logger = logging.getLogger(__name__)

JOKER_ENDPOINT = 'https://svc.joker.com/nic/replace'


class Authenticator(dns_common.DNSAuthenticator):
    """DNS Authenticator for Joker.

    This Authenticator uses the Joker DynDNS API to fulfill a dns-01 challenge.
    """

    description = 'Obtain certificates using a DNS TXT record (if you are using Joker for DNS).'
    ttl = 60

    def __init__(self, *args, **kwargs):
        super(Authenticator, self).__init__(*args, **kwargs)
        self.credentials = None

    @classmethod
    def add_parser_arguments(cls, add):  # pylint: disable=arguments-differ
        super(Authenticator, cls).add_parser_arguments(add, default_propagation_seconds=120)
        add('credentials', help='Joker credentials INI file.')

    def more_info(self):  # pylint: disable=missing-function-docstring
        return 'This plugin configures a DNS TXT record to respond to a dns-01 challenge using ' + \
               'the Joker v2 API.'

    def _setup_credentials(self):
        self.credentials = self._configure_credentials(
            'credentials',
            'Joker credentials INI file',
            {
                'username': 'domain-specific Joker dyndns username',
                'password': 'domain-specific Joker dyndns password',
                # 'domain': 'top-level domain for credentials',
            })

    def _perform(self, domain, validation_name, validation):
        self._get_joker_client(domain).add_txt_record(domain, validation_name, validation)

    def _cleanup(self, domain, validation_name, validation):
        self._get_joker_client(domain).del_txt_record(domain, validation_name, validation)

    def _get_joker_client(self, default_domain):
        username = self.credentials.conf('username')
        password = self.credentials.conf('password')
        domain = self.credentials.conf('domain')
        if not domain:
            domain = default_domain
        return _JokerClient(username, password, domain, self.ttl)


class _JokerClient(object):
    """
    Encapsulates all communication with the Joker.
    """

    # These are the error codes documented at https://help.dyn.com/remote-access-api/return-codes/
    error = {
       'badauth'  : 'Bad authorization (username or password)',
       'badsys'   : 'The system parameter given was not valid',

       'notfqdn'  : 'A Fully-Qualified Domain Name was not provided',
       'nohost'   : 'The hostname specified does not exist in the database',
       '!yours'   : 'The hostname specified exists, but not under the username currently being used',
       '!donator' : 'The offline setting was set, when the user is not a donator',
       '!active'  : 'The hostname specified is in a Custom DNS domain which has not yet been activated.',
       'abuse'    : 'The hostname specified is blocked for abuse; you should receive an email notification '
                     'which provides an unblock request link.  More info can be found on '
                     'https://www.dyndns.com/support/abuse.html',

       'numhost'  : 'System error: Too many or too few hosts found. Contact support@dyndns.org',
       'dnserr'   : 'System error: DNS error encountered. Contact support@dyndns.org',

       'nochg'    : 'No update required; unnecessary attempts to change to the current address are considered abusive',
    }

    def __init__(self, username, password, domain, ttl, endpoint=JOKER_ENDPOINT):
        self.endpoint = endpoint
        self.username = username
        self.password = password
        self.domain = domain
        self.ttl = ttl
        self.session = requests.Session()

    def add_txt_record(self, cert_domain, record_name, record_content):
        # Documentation for the Joker TXT record API is here:
        # https://joker.com/faq/content/6/496/en/let_s-encrypt-support.html

        # print(f'ADD domain:{cert_domain} record_name:{record_name} endpoint:{self.endpoint}')

        # Joker adds the domain to the end of the label of the TXT record that
        # it creates, but the record_name that certbot passed us already has
        # it so we need to remove it before calling the Joker API.
        dotdomain = '.' + self.domain
        if record_name.endswith(dotdomain):
            record_name = record_name[0:-len(dotdomain)]

        r = self.session.post(
            self.endpoint,
            data = {
                'username': self.username,
                'password': self.password,
                'zone': self.domain,
                'label': record_name,
                'type': 'TXT',
                'value': record_content,
                'ttl': self.ttl,
            })

        # print(f'ADD {r} {r.text}\n  REQ URL: {r.request.url}\n  REQ BODY: {r.request.body}\n')

        if r.status_code >= 300:
            self._handle_http_error(r.text, record_name, self.domain)

    def del_txt_record(self, domain, record_name, record_content):
        self.add_txt_record(domain, record_name, '')

    def _handle_http_error(self, error, record_name, domain_name):
        hint = self.error.get(error)
        raise errors.PluginError('Error setting {0} TXT record for {1}: {2}.{3}'
                                 .format(record_name, domain_name, error,
                                         ' ({0})'.format(hint) if hint else ''))
